api/filters/shared_instance_update_filter.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2018 The Service Manager Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"encoding/json"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/sirupsen/logrus"
+	"net/http"
+)
+
+const SharedInstanceUpdateFilterName = "SharedInstanceUpdateFilter"
+
+// SharingInstanceFilter validate the un/share request on an existing service instance
+type sharedInstanceUpdateFilter struct {
+	storageRepository storage.Repository
+}
+
+func NewSharedInstanceUpdateFilter(storageRepository storage.Repository) *sharedInstanceUpdateFilter {
+	return &sharedInstanceUpdateFilter{
+		storageRepository: storageRepository,
+	}
+}
+
+func (*sharedInstanceUpdateFilter) Name() string {
+	return SharedInstanceUpdateFilterName
+}
+
+func (sf *sharedInstanceUpdateFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	var reqServiceInstance types.ServiceInstance
+	err := util.BytesToObjectNoLabels(req.Body, &reqServiceInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceID := req.PathParams["resource_id"]
+	ctx := req.Context()
+
+	// Get instance from database
+	dbPersistedInstanceObject, err := storage.GetObjectByField(ctx, sf.storageRepository, types.ServiceInstanceType, "id", instanceID)
+	if err != nil {
+		return next.Handle(req)
+	}
+	persistedInstance := dbPersistedInstanceObject.(*types.ServiceInstance)
+
+	// Validate plan has not changed - we don't allow changing plan of shared instance, may have existing references
```

Comment:
why don't we allow that? if i have a shared DB and i want to change it from medium to large, why should i first delete all references? this could be a real problem..

Reply:
TODO
Adding a validation that the new plan `isShareable`.
Let Avi know that we need to warn the subaccount admin on a potential downtimes on its references.

Reply:
DONE
Added positive test - changing shared instance plan to a shareable plan
negative test - changing shared instance plan to a non-shareable plan
on both smaap and OSB flows

Reply:
@sigalmaya is handling the CLI changes
Avi updated.

Resolving